README.markdown
# README

RooLegacy implements read access for all spreadsheet types and read/write access for
Google spreadsheets. It can handle
* OpenOffice
* Excel
* Google spreadsheets
* Excelx
* LibreOffice
* CSV

### XLS

There is no support for formulas in RooLegacy for .xls files - you can get the result
of a formula but not the formula itself.

### Google Spreadsheet

Using RooLegacy to access Google spreadsheets requires you install the 'google-spreadsheet-ruby' gem separately.

## License

While RooLegacy is licensed under the MIT / Expat license, please note that the 'spreadsheet' gem [is released under](https://github.com/zdavatz/spreadsheet/blob/master/LICENSE.txt) the GPLv3 license.

## Usage:

```ruby
require 'roo'

s = RooLegacy::OpenOffice.new("myspreadsheet.ods")      # loads an OpenOffice Spreadsheet
s = RooLegacy::Excel.new("myspreadsheet.xls")           # loads an Excel Spreadsheet
s = RooLegacy::Google.new("myspreadsheetkey_at_google") # loads a Google Spreadsheet
s = RooLegacy::Excelx.new("myspreadsheet.xlsx")         # loads an Excel Spreadsheet for Excel .xlsx files
s = RooLegacy::CSV.new("mycsv.csv")                     # loads a CSV file

# You can use CSV to load TSV files, or files of a certain encoding by passing
# in options under the :csv_options key
s = RooLegacy::CSV.new("mytsv.tsv", csv_options: {col_sep: "\t"}) # TSV
s = RooLegacy::CSV.new("mycsv.csv", csv_options: {encoding: Encoding::ISO_8859_1}) # csv with explicit encoding

s.default_sheet = s.sheets.first             # first sheet in the spreadsheet file will be used

# s.sheets is an array which holds the names of the sheets within
# a spreadsheet.
# you can also write
# s.default_sheet = s.sheets[3] or
# s.default_sheet = 'Sheet 3'

s.cell(1,1)                                 # returns the content of the first row/first cell in the sheet
s.cell('A',1)                               # same cell
s.cell(1,'A')                               # same cell
s.cell(1,'A',s.sheets[0])                   # same cell

# almost all methods have an optional argument 'sheet'.
# If this parameter is omitted, the default_sheet will be used.

s.info                                      # prints infos about the spreadsheet file

s.first_row                                 # the number of the first row
s.last_row                                  # the number of the last row
s.first_column                              # the number of the first column
s.last_column                               # the number of the last column

# limited font information is available

s.font(1,1).bold?
s.font(1,1).italic?
s.font(1,1).underline?


# RooLegacy::Spreadsheet requires spreadsheet gem
require 'spreadsheet'

# Spreadsheet.open can accept both files and paths

xls = RooLegacy::Spreadsheet.open('./new_prices.xls')

# If the File.path or provided path string does not have an extension, you can optionally
# provide one as a string or symbol

xls = RooLegacy::Spreadsheet.open('./rails_temp_upload', extension: :xls)

# no more setting xls.default_sheet, just use this

xls.sheet('Info').row(1)
xls.sheet(0).row(1)

# excel likes to create random "Data01" sheets for macros
# use this to find the sheet with the most data to parse

xls.longest_sheet

# this excel file has multiple worksheets, let's iterate through each of them and process

xls.each_with_pagename do |name, sheet|
  p sheet.row(1)
end

# pull out a hash of exclusive column data (get rid of useless columns and save memory)

xls.each(:id => 'UPC',:qty => 'ATS') {|hash| arr << hash}
#=> hash will appear like {:upc=>727880013358, :qty => 12}

# NOTE: .parse does the same as .each, except it returns an array (similar to each vs. map)

# not sure exactly what a column will be named? try a wildcard search with the character *
# regex characters are allowed ('^price\s')
# case insensitive

xls.parse(:id => 'UPC*SKU',:qty => 'ATS*\sATP\s*QTY$')

# if you need to locate the header row and assign the header names themselves,
# use the :header_search option

xls.parse(:header_search => ['UPC*SKU','ATS*\sATP\s*QTY$'])
#=> each element will appear in this fashion:
#=> {"UPC" => 123456789012, "STYLE" => "987B0", "COLOR" => "blue", "QTY" => 78}

# want to strip out annoying unicode characters and surrounding white space?

xls.parse(:clean => true)

# another bonus feature is a patch to prevent the Spreadsheet gem from parsing
# thousands and thousands of blank lines. i got fed up after watching my computer
# nearly catch fire for 4 hours for a spreadsheet with only 200 ACTUAL lines
# - located in lib/roo/worksheet.rb
```


CHANGELOG
== unreleased

* enhancements
  * Extend RooLegacy::Spreadsheet.open to accept Tempfiles and other arguments responding to #path. Note they require an :extension option to be declared, as the tempfile mangles the extension. #84.

* bugfixes
  * Fix that paths with spaces in them would fail with URI::InvalidURIError. #121.

== 1.13.2 2013-12-23

* bugfixes
  * Fix that Excelx link-cells would blow up if the value wasn't a string. Due to the way Spreadsheet::Link is implemented the link text must be treated as a string. #92

== 1.13.1 2013-12-23

* bugfixes
  * Fix that Excelx creation could blow up due to nil rels files. #90

== 1.13.0 2013-12-05

* enhancements
  * Support extracting link data from Excel and Excelx spreadsheets,
    via Excel#read_cell(_content) and Excelx#hyperlink(?). #47
  * Support setting the Excel Spreadsheet mode via the :mode option. #88
  * Support Spreadsheet.open with a declared :extension that includes a leading '.'. #73
  * Enable file type detection for URI's with parameters / anchors. #51

* bugfixes
  * Fix that CSV#each_row could overwrite the filename when run against a uri. #77
  * Fix that #to_matrix wasn't respecting the sheet argument. #87

== 1.12.2 2013-09-11

* 1 enhancement
  * Support rubyzip >= 1.0.0. #65
  * Fix typo in deprecation notices. #63

== 1.12.1 2013-08-18

* 1 enhancement
  * Support :boolean fields for CSV export via #cell_to_csv. #59

* 2 bugfixes
  * Fix that Excelx would error on files with gaps in the numbering of their
  internal sheet#.xml files. #58
  * Fix that Base#info to preserve the original value of #default_sheet. #44

== 1.12.0 2013-08-18

* 3 deprecations
  * Rename Openoffice -> OpenOffice, Libreoffice -> LibreOffice, Csv -> CSV, and redirect the old names to the new constants
  * Enable RooLegacy::Excel, Excel2003XML, Excelx, OpenOffice to accept an options hash, and deprecate the old method argument based approach to supplying them options
  * RooLegacy's roo_rails_helper, aka the `spreadsheet` html-generating view method is currently deprecated with no replacement. If you find it helpful, tell http://github.com/Empact or extract it yourself.

* 9 enhancements
  * Add RooLegacy::Excelx#load_xml so that people can customize to their data, e.g. #23
  * Enable passing csv_options to RooLegacy::CSV, which are passed through to the underlying CSV call.
  * Enable passing options through from RooLegacy::Spreadsheet to any RooLegacy type.
  * Enable passing an :extension option to RooLegacy::Spreadsheet.new, which will override the extension detected on in the path #15
  * Switch from google-spreadsheet-ruby to google_drive for RooLegacy::Google access #40
  * Make all the classes consistent in that #read_cells is only effective if the sheet has not been read.
  * RooLegacy::Google supports login via oauth :access_token. #61
  * RooLegacy::Excel now exposes its Spreadsheet workbook via #workbook
  * Pull #load_xml down into RooLegacy::Base, and use it in Excel2003XML and OpenOffice.

* 2 changes
  * #formula? now returns truthy or falsey, rather than true/false.
  * Base#longest_sheet was moved to Excel, as it only worked under Excel

* 1 bugfix
  * Fix that RooLegacy::CSV#parse(headers: true) would blow up. #37

== 1.11.2 2013-04-10

* 4 bugfixes
  * Fix that RooLegacy::Spreadsheet.open wasn't tolerant to case differences.
  * Fix that RooLegacy::Excel2003XML loading was broken #27
  * Enable loading RooLegacy::Csv files from uris, just as other file types #31
  * Fix that Excelx "m/d/yy h:mm" was improperly being interpreted as date rather
    than datetime #29

== 1.11.1 2013-03-18

* 1 bugfix
  * Exclude test/log/roo.log test log file from the gemspec in order to avoid a
    rubygems warning: #26

== 1.11.0 2013-03-14

* 3 enhancements
  * Support ruby 2.0.0 by replacing Iconv with String#encode #19
  * Excelx: Loosen the format detection rules such that more are
    successfully detected #20
  * Delete the roo binary, which was useless and not declared in the gemspec

* 1 change
  * Drop support for ruby 1.8.x or lower. Required in order to easily support 2.0.0.

== 1.10.3 2013-03-03

* 1 bugfix
  * Support both nokogiri 1.5.5 and 1.5.6 (Karsten Richter) #18

* 1 enhancement
  * Relax our nokogiri dependency back to 1.4.0, as we have no particular reason
    to require a newer version.

== 1.10.2 2013-02-03

* 2 bugfixes
  * Support opening URIs with query strings https://github.com/Empact/roo/commit/abf94bdb59cabc16d4f7764025e88e3661983525
  * Support both http: & https: urls https://github.com/Empact/roo/commit/fc5c5899d96dd5f9fbb68125d0efc8ce9be2c7e1


lib/roo_legacy/base.rb
# encoding: utf-8

require 'tmpdir'
require 'stringio'

begin
  require 'zip/zipfilesystem'
  RooLegacy::ZipFile = Zip::ZipFile
rescue LoadError
  # For rubyzip >= 1.0.0
  require 'zip/filesystem'
  RooLegacy::ZipFile = Zip::File
end

# Base class for all other types of spreadsheets
class RooLegacy::Base
  include Enumerable

  TEMP_PREFIX = "oo_"
  LETTERS = %w{A B C D E F G H I J K L M N O P Q R S T U V W X Y Z}

  attr_reader :default_sheet, :headers

  # sets the line with attribute names (default: 1)
  attr_accessor :header_line

  class << self
    def split_coordinate(str)
      letter,number = split_coord(str)
      x = letter_to_number(letter)
      y = number
      return y, x
    end

    def split_coord(s)
      if s =~ /([a-zA-Z]+)([0-9]+)/
        letter = $1
        number = $2.to_i
      else
        raise ArgumentError
      end
      return letter, number
    end

    # convert a number to something like 'AB' (1 => 'A', 2 => 'B', ...)
    def number_to_letter(n)
      letters=""
      if n > 26
        while n % 26 == 0 && n != 0
          letters << 'Z'
          n = (n - 26) / 26
        end
        while n > 0
          num = n%26
          letters = LETTERS[num-1] + letters
          n = (n / 26)
        end
      else
        letters = LETTERS[n-1]
      end
      letters
    end

    # convert letters like 'AB' to a number ('A' => 1, 'B' => 2, ...)
    def letter_to_number(letters)
      result = 0
      while letters && letters.length > 0
        character = letters[0,1].upcase
        num = LETTERS.index(character)
        raise ArgumentError, "invalid column character '#{letters[0,1]}'" if num == nil
        num += 1
        result = result * 26 + num
        letters = letters[1..-1]
      end
      result
    end
  end

  def initialize(filename, options={}, file_warning=:error, tmpdir=nil)
    @filename = filename
    @options = options

    @cell = {}
    @cell_type = {}
    @cells_read = {}

    @first_row = {}
    @last_row = {}
    @first_column = {}
    @last_column = {}

    @header_line = 1
    @default_sheet = self.sheets.first
  end

  # sets the working sheet in the document
  # 'sheet' can be a number (1 = first sheet) or the name of a sheet.
  def default_sheet=(sheet)
    validate_sheet!(sheet)
    @default_sheet = sheet
    @first_row[sheet] = @last_row[sheet] = @first_column[sheet] = @last_column[sheet] = nil
    @cells_read[sheet] = false
  end

  # first non-empty column as a letter
  def first_column_as_letter(sheet=nil)
    self.class.number_to_letter(first_column(sheet))
  end

  # last non-empty column as a letter
  def last_column_as_letter(sheet=nil)
    self.class.number_to_letter(last_column(sheet))
  end

  # returns the number of the first non-empty row
  def first_row(sheet=nil)
    sheet ||= @default_sheet
    read_cells(sheet)
    @first_row[sheet] ||=
      begin
        impossible_value = 999_999 # more than a spreadsheet can hold
        result = impossible_value
        @cell[sheet].each_pair {|key,value|
          result = [result, key.first.to_i].min if value
        } if @cell[sheet]
        result unless result == impossible_value
      end
  end

  # returns the number of the last non-empty row
  def last_row(sheet=nil)
    sheet ||= @default_sheet
    read_cells(sheet)
    @last_row[sheet] ||=
      begin
        impossible_value = 0
        result = impossible_value
        @cell[sheet].each_pair {|key,value|
          result = [result, key.first.to_i].max if value
        } if @cell[sheet]
        result unless result == impossible_value
      end
  end

  # returns the number of the first non-empty column
  def first_column(sheet=nil)
    sheet ||= @default_sheet
    read_cells(sheet)
    @first_column[sheet] ||=
      begin
        impossible_value = 999_999 # more than a spreadsheet can hold
        result = impossible_value
        @cell[sheet].each_pair {|key,value|
          result = [result, key.last.to_i].min if value
        } if @cell[sheet]
        result unless result == impossible_value
      end
  end

  # returns the number of the last non-empty column
  def last_column(sheet=nil)
    sheet ||= @default_sheet
    read_cells(sheet)
    @last_column[sheet] ||=
      begin
        impossible_value = 0
        result = impossible_value
        @cell[sheet].each_pair {|key,value|
          result = [result, key.last.to_i].max if value
        } if @cell[sheet]
        result unless result == impossible_value
      end
  end

  # returns a rectangular area (default: all cells) as yaml-output
  # you can add additional attributes with the prefix parameter like:
  # oo.to_yaml({"file"=>"flightdata_2007-06-26", "sheet" => "1"})
  def to_yaml(prefix={}, from_row=nil, from_column=nil, to_row=nil, to_column=nil,sheet=nil)
    sheet ||= @default_sheet
    return '' unless first_row # empty result if there is no first_row in a sheet

    from_row ||= first_row(sheet)
    to_row ||= last_row(sheet)
    from_column ||= first_column(sheet)
    to_column ||= last_column(sheet)

    result = "--- \n"
    from_row.upto(to_row) do |row|
      from_column.upto(to_column) do |col|
        unless empty?(row,col,sheet)
          result << "cell_#{row}_#{col}: \n"
          prefix.each {|k,v|
            result << "  #{k}: #{v} \n"
          }
          result << "  row: #{row} \n"
          result << "  col: #{col} \n"
          result << "  celltype: #{self.celltype(row,col,sheet)} \n"
          value = cell(row,col,sheet)
          if celltype(row,col,sheet) == :time
            value = integer_to_timestring(value)
          end
          result << "  value: #{value} \n"
        end
      end
    end
    result
  end

  # write the current spreadsheet to stdout or into a file
  def to_csv(filename=nil,sheet=nil,separator=',')
    sheet ||= @default_sheet
    if filename
      File.open(filename,"w") do |file|
        write_csv_content(file,sheet,separator)
      end
      return true
    else
      sio = StringIO.new
      write_csv_content(sio,sheet,separator)
      sio.rewind
      return sio.read
    end
  end

  # returns a matrix object from the whole sheet or a rectangular area of a sheet
  def to_matrix(from_row=nil, from_column=nil, to_row=nil, to_column=nil,sheet=nil)
    require 'matrix'

    sheet ||= @default_sheet
    return Matrix.empty unless first_row

    from_row ||= first_row(sheet)
    to_row ||= last_row(sheet)
    from_column ||= first_column(sheet)
    to_column ||= last_column(sheet)

    Matrix.rows(from_row.upto(to_row).map do |row|
      from_column.upto(to_column).map do |col|
        cell(row,col,sheet)
      end
    end)
  end

  # call to_s method defined on subclasses
  def inspect
    to_s
  end

  # find a row either by row number or a condition
  # Caution: this works only within the default sheet -> set default_sheet before you call this method
  # (experimental. see examples in the test_roo.rb file)
  def find(*args) # :nodoc
    options = (args.last.is_a?(Hash) ? args.pop : {})

    case args[0]
    when Fixnum
      find_by_row(args[0])
    when :all
      find_by_conditions(options)
    else
      raise ArgumentError, "unexpected arg #{args[0].inspect}, pass a row index or :all"
    end
  end

  # returns all values in this row as an array
  # row numbers are 1,2,3,... like in the spreadsheet
  def row(rownumber,sheet=nil)
    sheet ||= @default_sheet
    read_cells(sheet)
    first_column(sheet).upto(last_column(sheet)).map do |col|
      cell(rownumber,col,sheet)
    end
  end

  # returns all values in this column as an array
  # column numbers are 1,2,3,... like in the spreadsheet
  def column(columnnumber,sheet=nil)
    if columnnumber.class == String
      columnnumber = RooLegacy::Excel.letter_to_number(columnnumber)
    end
    sheet ||= @default_sheet
    read_cells(sheet)
    first_row(sheet).upto(last_row(sheet)).map do |row|
      cell(row,columnnumber,sheet)
    end
  end

  # set a cell to a certain value
  # (this will not be saved back to the spreadsheet file!)
  def set(row,col,value,sheet=nil) #:nodoc:
    sheet ||= @default_sheet
    read_cells(sheet)
    row, col = normalize(row,col)
    cell_type =
      case value
      when Fixnum then :float
      when String, Float then :string
      else
        raise ArgumentError, "Type for #{value} not set"
      end

    set_value(row,col,value,sheet)
    set_type(row,col,cell_type,sheet)
  end

  # reopens and read a spreadsheet document
  def reload
    ds = @default_sheet
    reinitialize
    self.default_sheet = ds
  end

  # true if cell is empty
  def empty?(row, col, sheet=nil)
    sheet ||= @default_sheet
    read_cells(sheet)
    row,col = normalize(row,col)
    contents = cell(row, col, sheet)
    !contents || (celltype(row, col, sheet) == :string && contents.empty?) \
      || (row < first_row(sheet) || row > last_row(sheet) || col < first_column(sheet) || col > last_column(sheet))
  end

  # returns information of the spreadsheet document and all sheets within
  # this document.
  def info
    without_changing_default_sheet do
      result = "File: #{File.basename(@filename)}\n"+
        "Number of sheets: #{sheets.size}\n"+
        "Sheets: #{sheets.join(', ')}\n"
      n = 1
      sheets.each {|sheet|
        self.default_sheet = sheet
        result << "Sheet " + n.to_s + ":\n"
        unless first_row
          result << "  - empty -"
        else
          result << "  First row: #{first_row}\n"
          result << "  Last row: #{last_row}\n"
          result << "  First column: #{self.class.number_to_letter(first_column)}\n"
          result << "  Last column: #{self.class.number_to_letter(last_column)}"
        end
        result << "\n" if sheet != sheets.last
        n += 1
      }
      result
    end
  end

  # returns an XML representation of all sheets of a spreadsheet file
  def to_xml
    Nokogiri::XML::Builder.new do |xml|
      xml.spreadsheet {
        self.sheets.each do |sheet|
          self.default_sheet = sheet
          xml.sheet(:name => sheet) { |x|
            if first_row and last_row and first_column and last_column
              # sonst gibt es Fehler bei leeren Blaettern
              first_row.upto(last_row) do |row|
                first_column.upto(last_column) do |col|
                  unless empty?(row,col)
                    x.cell(cell(row,col),
                      :row =>row,
                      :column => col,
                      :type => celltype(row,col))
                  end
                end
              end
            end
          }
        end
      }
    end.to_xml
  end

  # when a method like spreadsheet.a42 is called
  # convert it to a call of spreadsheet.cell('a',42)
  def method_missing(m, *args)
    # #aa42 => #cell('aa',42)
    # #aa42('Sheet1')  => #cell('aa',42,'Sheet1')
    if m =~ /^([a-z]+)(\d)$/
      col = self.class.letter_to_number($1)
      row = $2.to_i
      if args.empty?
        cell(row,col)
      else
        cell(row,col,args.first)
      end
    else
      super
    end
  end

  # access different worksheets by calling spreadsheet.sheet(1)
  # or spreadsheet.sheet('SHEETNAME')
  def sheet(index,name=false)
    @default_sheet = String === index ? index : self.sheets[index]
    name ? [@default_sheet,self] : self
  end

  # iterate through all worksheets of a document
  def each_with_pagename
    self.sheets.each do |s|
      yield sheet(s,true)
    end
  end

  # by passing in headers as options, this method returns
  # specific columns from your header assignment
  # for example:
  # xls.sheet('New Prices').parse(:upc => 'UPC', :price => 'Price') would return:
  # [{:upc => 123456789012, :price => 35.42},..]

  # the queries are matched with regex, so regex options can be passed in
  # such as :price => '^(Cost|Price)'
  # case insensitive by default


  # by using the :header_search option, you can query for headers
  # and return a hash of every row with the keys set to the header result
  # for example:
  # xls.sheet('New Prices').parse(:header_search => ['UPC*SKU','^Price*\sCost\s'])

  # that example searches for a column titled either UPC or SKU and another
  # column titled either Price or Cost (regex characters allowed)
  # * is the wildcard character

  # you can also pass in a :clean => true option to strip the sheet of
  # odd unicode characters and white spaces around columns

  def each(options={})
    if options.empty?
      1.upto(last_row) do |line|
        yield row(line)
      end
    else
      if options[:clean]
        options.delete(:clean)
        @cleaned ||= {}
        @cleaned[@default_sheet] || clean_sheet(@default_sheet)
      end

      if options[:header_search]
        @headers = nil
        @header_line = row_with(options[:header_search])
      elsif [:first_row,true].include?(options[:headers])
        @headers = []
        row(first_row).each_with_index {|x,i| @headers << [x,i + 1]}
      else
        set_headers(options)
      end

      headers = @headers ||
        Hash[(first_column..last_column).map do |col|
          [cell(@header_line,col), col]
        end]

      @header_line.upto(last_row) do |line|
        yield(Hash[headers.map {|k,v| [k,cell(line,v)]}])
      end
    end
  end

  def parse(options={})
    ary = []
    if block_given?
      each(options) {|row| ary << yield(row)}
    else
      each(options) {|row| ary << row}
    end
    ary
  end

  def row_with(query,return_headers=false)
    query.map! {|x| Array(x.split('*'))}
    line_no = 0
    each do |row|
      line_no += 1
      # makes sure headers is the first part of wildcard search for priority
      # ex. if UPC and SKU exist for UPC*SKU search, UPC takes the cake
      headers = query.map do |q|
        q.map {|i| row.grep(/#{i}/i)[0]}.compact[0]
      end.compact

      if headers.length == query.length
        @header_line = line_no
        return return_headers ? headers : line_no
      elsif line_no > 100
        raise "Couldn't find header row."
      end
    end
  end

  protected

  def load_xml(path)
    ::File.open(path, 'rb') do |file|
      ::Nokogiri::XML(file)
    end
  end

  def file_type_check(filename, ext, name, warning_level, packed=nil)
    new_expression = {
      '.ods' => 'RooLegacy::OpenOffice.new',
      '.xls' => 'RooLegacy::Excel.new',
      '.xlsx' => 'RooLegacy::Excelx.new',
      '.csv' => 'RooLegacy::CSV.new',
      '.xml' => 'RooLegacy::Excel2003XML.new',
    }
    if packed == :zip
	    # lalala.ods.zip => lalala.ods
	    # hier wird KEIN unzip gemacht, sondern nur der Name der Datei
	    # getestet, falls es eine gepackte Datei ist.
	    filename = File.basename(filename,File.extname(filename))
    end
    case ext
    when '.ods', '.xls', '.xlsx', '.csv', '.xml'
      correct_class = "use #{new_expression[ext]} to handle #{ext} spreadsheet files. This has #{File.extname(filename).downcase}"
    else
      raise "unknown file type: #{ext}"
    end

    if uri?(filename) && qs_begin = filename.rindex('?')
      filename = filename[0..qs_begin-1]
    end
    if File.extname(filename).downcase != ext
      case warning_level
      when :error
        warn correct_class
        raise TypeError, "#{filename} is not #{name} file"
      when :warning
        warn "are you sure, this is #{name} spreadsheet file?"
        warn correct_class
      when :ignore
        # ignore
      else
        raise "#{warning_level} illegal state of file_warning"
      end
    end
  end

  # konvertiert einen Key in der Form "12,45" (=row,column) in
  # ein Array mit numerischen Werten ([12,45])
  # Diese Methode ist eine temp. Loesung, um zu erforschen, ob der
  # Zugriff mit numerischen Keys schneller ist.
  def key_to_num(str)
    r,c = str.split(',')
    [r.to_i,c.to_i]
  end

  # see: key_to_num
  def key_to_string(arr)
    "#{arr[0]},#{arr[1]}"
  end

  private

  def find_by_row(row_index)
    row_index += (header_line - 1) if @header_line

    row(row_index).size.times.map do |cell_index|
      cell(row_index, cell_index + 1)
    end
  end

  def find_by_conditions(options)
    rows = first_row.upto(last_row)
    header_for = Hash[1.upto(last_column).map do |col|
      [col, cell(@header_line,col)]
    end]

    # are all conditions met?
    conditions = options[:conditions]
    if conditions && !conditions.empty?
      column_with = header_for.invert
      rows = rows.select do |i|
        conditions.all? { |key,val| cell(i,column_with[key]) == val }
      end
    end

    if options[:array]
      rows.map {|i| self.row(i) }
    else
      rows.map do |i|
        Hash[1.upto(self.row(i).size).map do |j|
          [header_for.fetch(j), cell(i,j)]
        end]
      end
    end
  end


  def without_changing_default_sheet
    original_default_sheet = default_sheet
    yield
  ensure
    self.default_sheet = original_default_sheet
  end

  def reinitialize
    initialize(@filename)
  end

  def make_tmpdir(tmp_root = nil)
    Dir.mktmpdir(TEMP_PREFIX, tmp_root || ENV['ROO_TMP']) do |tmpdir|
      yield tmpdir
    end
  end

  def clean_sheet(sheet)
    read_cells(sheet)
    @cell[sheet].each_pair do |coord,value|
      if String === value
        @cell[sheet][coord] = sanitize_value(value)
      end
    end
    @cleaned[sheet] = true
  end

  def sanitize_value(v)
    v.unpack('U*').select {|b| b < 127}.pack('U*').strip
  end

  def set_headers(hash={})
    # try to find header row with all values or give an error
    # then create new hash by indexing strings and keeping integers for header array
    @headers = row_with(hash.values,true)
    @headers = Hash[hash.keys.zip(@headers.map {|x| header_index(x)})]
  end

  def header_index(query)
    row(@header_line).index(query) + first_column
  end

  def set_value(row,col,value,sheet=nil)
    sheet ||= @default_sheet
    @cell[sheet][[row,col]] = value
  end

  def set_type(row,col,type,sheet=nil)
    sheet ||= @default_sheet
    @cell_type[sheet][[row,col]] = type
  end

  # converts cell coordinate to numeric values of row,col
  def normalize(row,col)
    if row.class == String
      if col.class == Fixnum
        # ('A',1):
        # ('B', 5) -> (5, 2)
        row, col = col, row
      else
        raise ArgumentError
      end
    end
    if col.class == String
      col = self.class.letter_to_number(col)
    end
    return row,col
  end

  def uri?(filename)
    begin
      filename.start_with?("http://", "https://")
    rescue
      false
    end
  end

  def download_uri(uri, tmpdir)
    require 'open-uri'
    tempfilename = File.join(tmpdir, File.basename(uri))
    response = ''
    begin
      File.open(tempfilename,"wb") do |file|
        open(uri, "User-Agent" => "Ruby/#{RUBY_VERSION}") { |net|
          file.write(net.read)
        }
      end
    rescue OpenURI::HTTPError
      raise "could not open #{uri}"
    end
    tempfilename
  end

  def open_from_stream(stream, tmpdir)
    tempfilename = File.join(tmpdir, "spreadsheet")
    File.open(tempfilename,"wb") do |file|
      file.write(stream[7..-1])
    end
    File.join(tmpdir, "spreadsheet")
  end

  def unzip(filename, tmpdir)
    RooLegacy::ZipFile.open(filename) do |zip|
      process_zipfile_packed(zip, tmpdir)
    end
  end

  # check if default_sheet was set and exists in sheets-array
  def validate_sheet!(sheet)
    case sheet
    when nil
      raise ArgumentError, "Error: sheet 'nil' not valid"
    when Fixnum
      self.sheets.fetch(sheet-1) do
        raise RangeError, "sheet index #{sheet} not found"
      end
    when String
      if !sheets.include? sheet
        raise RangeError, "sheet '#{sheet}' not found"
      end
    else
      raise TypeError, "not a valid sheet type: #{sheet.inspect}"
    end
  end

  def process_zipfile_packed(zip, tmpdir, path='')
    if zip.file.file? path
      # extract and return filename
      File.open(File.join(tmpdir, path),"wb") do |file|
        file.write(zip.read(path))
      end
      File.join(tmpdir, path)
    else
      ret=nil
      path += '/' unless path.empty?
      zip.dir.foreach(path) do |filename|
        ret = process_zipfile_packed(zip, tmpdir, path + filename)
      end
      ret
    end
  end

  # Write all cells to the csv file. File can be a filename or nil. If the this
  # parameter is nil the output goes to STDOUT
  def write_csv_content(file=nil,sheet=nil,separator=',')
    file ||= STDOUT
    if first_row(sheet) # sheet is not empty
      1.upto(last_row(sheet)) do |row|
        1.upto(last_column(sheet)) do |col|
          file.print(separator) if col > 1
          file.print cell_to_csv(row,col,sheet)
        end
        file.print("\n")
      end # sheet not empty
    end
  end

  # The content of a cell in the csv output
  def cell_to_csv(row, col, sheet)
    if empty?(row,col,sheet)
      ''
    else
      onecell = cell(row,col,sheet)

      case celltype(row,col,sheet)
      when :string
        unless onecell.empty?
          %{"#{onecell.gsub(/"/,'""')}"}
        end
      when :boolean
        %{"#{onecell.gsub(/"/,'""').downcase}"}
      when :float, :percentage
        if onecell == onecell.to_i
          onecell.to_i.to_s
        else
          onecell.to_s
        end
      when :formula
        case onecell
        when String
          unless onecell.empty?
            %{"#{onecell.gsub(/"/,'""')}"}
          end
        when Float
          if onecell == onecell.to_i
            onecell.to_i.to_s
          else
            onecell.to_s
          end
        when DateTime
          onecell.to_s
        else
          raise "unhandled onecell-class #{onecell.class}"
        end
      when :date, :datetime
        onecell.to_s
      when :time
        integer_to_timestring(onecell)
      when :link
          %{"#{onecell.url.gsub(/"/,'""')}"}
      else
        raise "unhandled celltype #{celltype(row,col,sheet)}"
      end || ""
    end
  end

  private

  # converts an integer value to a time string like '02:05:06'
  def integer_to_timestring(content)
    h = (content/3600.0).floor
    content = content - h*3600
    m = (content/60.0).floor
    content = content - m*60
    s = content
    sprintf("%02d:%02d:%02d",h,m,s)
  end
end
